File: forms/bone-marrow-donation-assessment/back-end-with-loco/tests/engine/donor_grader_test.rs
use bone_marrow_donation_assessment_loco_crate::engine::donor_grader::{evaluate_rules, grade};
use bone_marrow_donation_assessment_loco_crate::engine::donor_rules::all_rules;
use bone_marrow_donation_assessment_loco_crate::engine::types::*;

/// Build a clean donor with ideal HLA match, ASA I, normal labs.
fn ideal_donor() -> AssessmentData {
    AssessmentData {
        demographics: Demographics {
            first_name: "Alex".to_string(),
            last_name: "Donor".to_string(),
            date_of_birth: "1990-04-11".to_string(),
            sex: "female".to_string(),
            weight: Some(72.0),
            height: Some(170.0),
            bmi: Some(24.9),
        },
        donor_registration_hla_typing: DonorRegistrationHlaTyping {
            donor_registry: "NHS Blood and Transplant".to_string(),
            donor_registry_id: "REG-001".to_string(),
            registration_date: "2024-01-01".to_string(),
            donation_type: "unrelated".to_string(),
            recipient_relationship: String::new(),
            hla_a: "A*01:01".to_string(),
            hla_b: "B*08:01".to_string(),
            hla_c: "C*07:01".to_string(),
            hla_drb1: "DRB1*03:01".to_string(),
            hla_dqb1: "DQB1*02:01".to_string(),
            hla_dpb1: "DPB1*04:01".to_string(),
            hla_match_level: "10-of-10".to_string(),
            crossmatch_result: "negative".to_string(),
            previous_donation: "no".to_string(),
            previous_donation_details: String::new(),
        },
        medical_history: MedicalHistory::default(),
        physical_examination: PhysicalExamination {
            oxygen_saturation: Some(99),
            general_appearance: "well".to_string(),
            cardiovascular_examination: "normal".to_string(),
            respiratory_examination: "normal".to_string(),
            abdominal_examination: "normal".to_string(),
            venous_access_assessment: "good".to_string(),
            posterior_iliac_crest_assessment: "suitable".to_string(),
            ..Default::default()
        },
        haematological_assessment: HaematologicalAssessment {
            haemoglobin: Some(14.0),
            platelet_count: Some(250.0),
            creatinine: Some(80.0),
            coagulation_screen: "normal".to_string(),
            liver_function: "normal".to_string(),
            ..Default::default()
        },
        infectious_disease_screening: InfectiousDiseaseScreening {
            hiv_status: "negative".to_string(),
            hepatitis_b_surface_antigen: "negative".to_string(),
            hepatitis_b_core_antibody: "negative".to_string(),
            hepatitis_c_abntibody: "negative".to_string(),
            htlv_status: "negative".to_string(),
            syphilis_screen: "negative".to_string(),
            tuberculosis_screen: "negative".to_string(),
            vaccination_up_to_date: "yes".to_string(),
            ..Default::default()
        },
        anaesthetic_assessment: AnaestheticAssessment {
            asa_grade: "I".to_string(),
            mallampati_score: "I".to_string(),
            smoking_status: "never".to_string(),
            alcohol_use: "none".to_string(),
            ..Default::default()
        },
        collection_method_assessment: CollectionMethodAssessment {
            preferred_method: "pbsc".to_string(),
            final_collection_method: "pbsc".to_string(),
            gcsf_eligible: "yes".to_string(),
            venous_access_suitable_for_apheresis: "yes".to_string(),
            central_line_required: "no".to_string(),
            ..Default::default()
        },
        psychological_readiness: PsychologicalReadiness {
            understands_procedure: "yes".to_string(),
            understands_risks: "yes".to_string(),
            voluntary_decision: "yes".to_string(),
            coercion_concerns: "no".to_string(),
            anxiety_about_procedure: "none".to_string(),
            support_network: "yes".to_string(),
            donor_advocate_consulted: "yes".to_string(),
            willing_to_proceed: "yes".to_string(),
            ..Default::default()
        },
        consent_eligibility: ConsentEligibility {
            informed_consent_given: "yes".to_string(),
            consent_form_signed: "yes".to_string(),
            information_leaflet_provided: "yes".to_string(),
            questions_answered: "yes".to_string(),
            ..Default::default()
        },
    }
}

#[test]
fn ideal_donor_is_suitable_low_risk() {
    let data = ideal_donor();
    let result = grade(&data);
    assert_eq!(result.eligibility, "suitable");
    assert_eq!(result.overall_risk, "low");
    // HLA-001 (ideal-donor marker) is allowed to fire but should not penalise.
    assert!(result.fired_rules.iter().any(|r| r.id == "HLA-001"));
    assert!(result.fired_rules.iter().any(|r| r.id == "AN-001"));
}

#[test]
fn hiv_positive_is_unsuitable_critical() {
    let mut data = ideal_donor();
    data.infectious_disease_screening.hiv_status = "positive".to_string();
    let result = grade(&data);
    assert_eq!(result.eligibility, "unsuitable");
    assert_eq!(result.overall_risk, "critical");
    assert!(result.fired_rules.iter().any(|r| r.id == "ID-001"));
}

#[test]
fn seven_of_ten_match_fires_hla_004() {
    let mut data = ideal_donor();
    data.donor_registration_hla_typing.hla_match_level = "7-of-10".to_string();
    let fired = evaluate_rules(&data);
    assert!(fired.iter().any(|r| r.id == "HLA-004" && r.grade == 4));
}

#[test]
fn single_grade2_finding_is_conditionally_suitable_moderate() {
    let mut data = ideal_donor();
    data.haematological_assessment.haemoglobin = Some(11.5);
    let result = grade(&data);
    assert_eq!(result.eligibility, "conditionally-suitable");
    assert_eq!(result.overall_risk, "moderate");
}

#[test]
fn assessor_override_wins() {
    let mut data = ideal_donor();
    // engine would say suitable; assessor records unsuitable.
    data.consent_eligibility.eligibility_decision = "unsuitable".to_string();
    let result = grade(&data);
    assert_eq!(result.eligibility, "unsuitable");
    // suggested still reflects engine output (suitable).
    assert_eq!(result.suggested_eligibility, "suitable");
}

#[test]
fn rule_ids_are_unique() {
    let rules = all_rules();
    let mut ids: Vec<&str> = rules.iter().map(|r| r.id).collect();
    ids.sort();
    let len_before = ids.len();
    ids.dedup();
    assert_eq!(ids.len(), len_before, "rule IDs must be unique");
}

File: forms/bone-marrow-donation-assessment/front-end-with-svelte/src/lib/engine/donor-grader.test.ts
import { describe, it, expect } from 'vitest';
import { calculateDonorGrade } from './donor-grader';
import type { AssessmentData } from './types';

/** A blank assessment built inline (no store import — keeps vitest free of $app). */
function base(): AssessmentData {
	return {
		demographics: { firstName: '', lastName: '', dateOfBirth: '', sex: '', weight: null, height: null, bmi: null },
		donorRegistrationHlaTyping: {
			donorRegistry: '', donorRegistryId: '', registrationDate: '', donationType: '', recipientRelationship: '',
			hlaA: '', hlaB: '', hlaC: '', hlaDrb1: '', hlaDqb1: '', hlaDpb1: '', hlaMatchLevel: '',
			crossmatchResult: '', previousDonation: '', previousDonationDetails: ''
		},
		medicalHistory: {
			hasAutoimmuneDisease: '', autoimmuneDetails: '', hasMalignancy: '', malignancyDetails: '',
			hasCardiovascularDisease: '', cardiovascularDetails: '', hasRespiratoryDisease: '', respiratoryDetails: '',
			hasRenalDisease: '', renalDetails: '', hasHepaticDisease: '', hepaticDetails: '', hasBleedingDisorder: '',
			bleedingDisorderDetails: '', hasNeurologicalCondition: '', neurologicalDetails: '', currentMedications: '',
			drugAllergies: '', previousSurgery: '', surgeryDetails: ''
		},
		physicalExamination: {
			bpSystolic: null, bpDiastolic: null, heartRate: null, temperature: null, respiratoryRate: null,
			oxygenSaturation: null, generalAppearance: '', cardiovascularExamination: '', cardiovascularFindings: '',
			respiratoryExamination: '', respiratoryFindings: '', abdominalExamination: '', abdominalFindings: '',
			venousAccessAssessment: '', posteriorIliacCrestAssessment: ''
		},
		haematologicalAssessment: {
			haemoglobin: null, whiteCellCount: null, plateletCount: null, neutrophilCount: null, lymphocyteCount: null,
			haematocrit: null, mcv: null, bloodGroup: '', coagulationScreen: '', coagulationDetails: '', ferritin: null,
			creatinine: null, liverFunction: '', liverFunctionDetails: ''
		},
		infectiousDiseaseScreening: {
			hivStatus: '', hepatitisBSurfaceAntigen: '', hepatitisBCoreAntibody: '', hepatitisCAbntibody: '', htlvStatus: '',
			syphilisScreen: '', cmvStatus: '', ebvStatus: '', toxoplasmaStatus: '', tuberculosisScreen: '', recentTravel: '',
			travelDetails: '', recentInfection: '', infectionDetails: '', vaccinationUpToDate: ''
		},
		anaestheticAssessment: {
			asaGrade: '', previousAnaesthetic: '', anaestheticComplications: '', complicationDetails: '',
			familyAnaestheticProblems: '', familyProblemDetails: '', mallampatiScore: '', airwayConcerns: '', airwayDetails: '',
			nilByMouthConfirmed: '', smokingStatus: '', alcoholUse: '', anaestheticPlan: ''
		},
		collectionMethodAssessment: {
			preferredMethod: '', recipientPreference: '', finalCollectionMethod: '', gcsfEligible: '', gcsfContraindications: '',
			venousAccessSuitableForApheresis: '', centralLineRequired: '', estimatedDonorWeightKg: null, targetCd34Dose: null,
			estimatedCollectionDays: null, boneMarrowHarvestVolumeMl: null, autologousBloodDonation: ''
		},
		psychologicalReadiness: {
			understandsProcedure: '', understandsRisks: '', voluntaryDecision: '', coercionConcerns: '', coercionDetails: '',
			anxietyAboutProcedure: '', previousPsychologicalIssues: '', psychologicalIssueDetails: '', supportNetwork: '',
			timeOffWorkArranged: '', donorAdvocateConsulted: '', willingToProceed: ''
		},
		consentEligibility: {
			informedConsentGiven: '', consentFormSigned: '', consentDate: '', witnessName: '', witnessRole: '',
			informationLeafletProvided: '', questionsAnswered: '', eligibilityDecision: '', eligibilityConditions: '',
			deferralReason: '', deferralDuration: '', assessorName: '', assessorRole: '', assessmentDate: ''
		}
	};
}

describe('calculateDonorGrade', () => {
	it('grades an ideal donor as suitable / low risk', () => {
		const d = base();
		d.donorRegistrationHlaTyping.hlaMatchLevel = '10-of-10';
		d.anaestheticAssessment.asaGrade = 'I';
		const g = calculateDonorGrade(d);
		expect(g.overallRisk).toBe('low');
		expect(g.eligibility).toBe('suitable');
	});

	it('fires HLA and anaesthetic rules and derives moderate risk for a 9/10 ASA II donor', () => {
		const d = base();
		d.donorRegistrationHlaTyping.hlaMatchLevel = '9-of-10';
		d.anaestheticAssessment.asaGrade = 'II';
		const g = calculateDonorGrade(d);
		expect(g.firedRules.some((r) => r.id === 'HLA-002')).toBe(true);
		expect(g.overallRisk).toBe('moderate');
		expect(g.eligibility).toBe('conditionally-suitable');
	});

	it('grades a critical contraindication as critical risk / unsuitable', () => {
		const d = base();
		d.infectiousDiseaseScreening.hivStatus = 'positive';
		const g = calculateDonorGrade(d);
		expect(g.overallRisk).toBe('critical');
		expect(g.eligibility).toBe('unsuitable');
		expect(g.firedRules.some((r) => r.id === 'ID-001')).toBe(true);
	});

	it('honours an explicit eligibility decision over the derived value', () => {
		const d = base();
		d.donorRegistrationHlaTyping.hlaMatchLevel = '10-of-10';
		d.consentEligibility.eligibilityDecision = 'unsuitable';
		const g = calculateDonorGrade(d);
		expect(g.eligibility).toBe('unsuitable');
	});

	it('reports a positive crossmatch as a high-priority flag', () => {
		const d = base();
		d.donorRegistrationHlaTyping.crossmatchResult = 'positive';
		const g = calculateDonorGrade(d);
		expect(g.additionalFlags.some((f) => f.id === 'FLAG-XM-001' && f.priority === 'high')).toBe(true);
	});

	it('exposes the HLA match and collection method labels', () => {
		const d = base();
		d.donorRegistrationHlaTyping.hlaMatchLevel = '8-of-10';
		d.collectionMethodAssessment.finalCollectionMethod = 'bone-marrow';
		const g = calculateDonorGrade(d);
		expect(g.hlaMatchLevel).toContain('8/10');
		expect(g.collectionMethod).toContain('Bone Marrow');
	});
});
